### PR TITLE
Fix serializer check to permit fast-path serialization logic to run

### LINKED ISSUE #53814: Fix serializer check to permit fast-path serialization logic to run
https://github.com/dotnet/aspnetcore/pull/33296#issuecomment-856034647

### PR DESCRIPTION
**Pull request by layomia**
Fixes https://github.com/dotnet/runtime/issues/53814.

FYI @pranavkm @eerhardt

### CODE REVIEW

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerContext.cs [hunk 1]
```diff
@@ -37,9 +37,12 @@ internal bool CanUseSerializationLogic
                             // Guard against unsupported features
                             Options.Converters.Count == 0 &&
                             Options.Encoder == null &&
-                            Options.NumberHandling == JsonNumberHandling.Strict &&
+                            (((JsonNumberHandling.WriteAsString | JsonNumberHandling.AllowNamedFloatingPointLiterals) & Options.NumberHandling) == 0) &&
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerContext.cs:40]: If I understand the bug correctly, it is us miscalculating `_canUserSerializationLogic` in cases where we have `Options.NumberHandling == JsonNumberHandling.AllowReadingFromString`?

Also, nit:
```suggestion
                            ((Options.NumberHandling & (JsonNumberHandling.WriteAsString | JsonNumberHandling.AllowNamedFloatingPointLiterals)) == 0) &&
```

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerContext.cs [hunk 2]
```diff
@@ -37,9 +37,12 @@ internal bool CanUseSerializationLogic
...
                             Options.ReferenceHandlingStrategy == ReferenceHandlingStrategy.None &&
-                            // Ensure options values are consistent with expected defaults.
+#pragma warning disable SYSLIB0020
+                            !Options.IgnoreNullValues && // This property is obsolete.
+#pragma warning restore SYSLIB0020
+                                  // Ensure options values are consistent with expected defaults.
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerContext.cs:45]: Looks like indentation is off for this comment?

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerContext.cs (hunk 1 above)
Comment (layomia) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerContext.cs:40]: Yes `AllowReadingFromString` has no bearing on serialization, so we can allow fast-path serialize logic to run when it is set.

**Review by eiriktsarpalis — APPROVED** [review]

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerContext.cs [hunk 3]
```diff
@@ -37,8 +37,11 @@ internal bool CanUseSerializationLogic
                             // Guard against unsupported features
                             Options.Converters.Count == 0 &&
                             Options.Encoder == null &&
-                            Options.NumberHandling == JsonNumberHandling.Strict &&
+                            (Options.NumberHandling & (JsonNumberHandling.WriteAsString | JsonNumberHandling.AllowNamedFloatingPointLiterals)) == 0 &&
```
Comment (eerhardt) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerContext.cs:40]: Why don't we support `NumberHandling == JsonNumberHandling.Strict` anymore?

Comment (eerhardt) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerContext.cs:40]: Oh, I missed the `== 0` part at the end. Maybe it would be good to add a comment here what the policy is?

File: src/libraries/System.Text.Json/tests/System.Text.Json.SourceGeneration.Tests/SerializationLogicTests.cs [hunk 4]
```diff
@@ -39,23 +58,51 @@ public static void DictionaryFastPathPrimitiveValueSupported()
...
+        // Options with features that aren't supported in the generated serialization funcs.
         public static IEnumerable<object[]> GetOptionsUsingUnsupportedFeatures()
         {
-            yield return new object[] { new JsonSerializerOptions { Converters = { new JsonStringEnumConverter() } } };
-            yield return new object[] { new JsonSerializerOptions { Encoder = JavaScriptEncoder.UnsafeRelaxedJsonEscaping } };
-            yield return new object[] { new JsonSerializerOptions { NumberHandling = JsonNumberHandling.WriteAsString } };
-            yield return new object[] { new JsonSerializerOptions { ReferenceHandler = ReferenceHandler.Preserve } };
-            yield return new object[] { new JsonSerializerOptions { ReferenceHandler = ReferenceHandler.IgnoreCycles } };
-            yield return new object[] { new JsonSerializerOptions { ReferenceHandler = ReferenceHandler.IgnoreCycles } };
+            yield return new object[] { new JsonSerializerOptions(s_compatibleOptions) { Converters = { new JsonStringEnumConverter() } } };
```
Comment (eerhardt) [inline src/libraries/System.Text.Json/tests/System.Text.Json.SourceGeneration.Tests/SerializationLogicTests.cs:82]: Why the change to the existing tests here? Do we no longer care about testing the scenarios that don't have `s_compatibleOptions`?

Would it be better to leave the existing tests and add these new ones as additional new cases?

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerContext.cs (hunk 3 above)
Comment (layomia) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerContext.cs:40]: Yup, added a comment.

File: src/libraries/System.Text.Json/tests/System.Text.Json.SourceGeneration.Tests/SerializationLogicTests.cs (hunk 4 above)
Comment (layomia) [reply src/libraries/System.Text.Json/tests/System.Text.Json.SourceGeneration.Tests/SerializationLogicTests.cs:84]: I'll move them to `GetIncompatibleOptions` below since the more pertinent issue with them (in my view) is that they don't set these two properties - https://github.com/dotnet/runtime/blob/552b8678611afb08e4743e122b917847dae6e2ce/src/libraries/System.Text.Json/tests/System.Text.Json.SourceGeneration.Tests/ContextClasses.cs#L35-L36.
